### PR TITLE
librbd: support the ability to process parent data prior to copyup

### PR DESCRIPTION
**Pull request by dillaman**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by orozery** [comment]
Just wondering:
The prepare_copyup function will get the parent data minus places where there's overlapping writes, correct?
How hard will it be to instead build and pass-on a merged representation of parent data plus new writes?

**Comment by dillaman** [comment]
> Just wondering:
> The prepare_copyup function will get the parent data minus places where there's overlapping writes, correct?

Correct. There are two cases for a simple (standard) copyup. (1) there are already snapshots on the image or it's a copy-on-read operation, so the parent data will not have the overlap subtracted because the overlapping writes don't exist yet (at snapshot zero vs the HEAD revision), or (2) the copyup is destined for the HEAD revision so it subtracts any parent data that is about to be overwritten.

> How hard will it be to instead build and pass-on a merged representation of parent data plus new writes?

The new writes would have already been encrypted, correct? Whereas the parent data (passed via the process hook) is unencrypted. The new writes should have also already been properly block aligned, so any unaligned parent data extents you receive can be assumed to be zeroed. 

Now, there is one caveat to all of this and that will occur in the deep-copy `ObjectCopyRequest` state machine where we will potentially pass through the hook the parent data deltas at different points in snapshot history. These won't have any HEAD revision overlapping writes subtracted from them, but it does bring up the question of how to best allow the crypto hook to properly process potentially unaligned snapshot deltas or worse -- discarded / zeroed extents.

**Comment by dillaman** [comment]
> These won't have any HEAD revision overlapping writes subtracted from them, but it does bring up the question of how to best allow the crypto hook to properly process potentially unaligned snapshot deltas or worse -- discarded / zeroed extents.

I tweaked the hook a little bit to pass a new `SparseExtents` type which allows a mapping between an extent and a state (i.e. `SPARSE_EXTENT_STATE_DATA` vs `SPARSE_EXTENT_STATE_ZEROED`). That will allow writes to remain sparse if possible or allow the crypto layer to change some `ZEROED` extents to encrypted `DATA` extents.

File: src/librbd/io/CopyupRequest.cc [hunk 1]
```diff
@@ -401,7 +402,7 @@ void CopyupRequest<I>::copyup() {
 
   ldout(cct, 20) << dendl;
 
-  bool copy_on_read = m_pending_requests.empty();
+  bool copy_on_read = m_pending_requests.empty() && !m_deep_copied;
```
Comment (trociny) [inline src/librbd/io/CopyupRequest.cc:398]: @dillaman Just wondering, couldn't `is_deep_copy()` be used instead of introducing a new variable?

Comment (dillaman) [reply src/librbd/io/CopyupRequest.cc:398]: I wanted to avoid any possibility of a race since if the `migration_info` header changes but we already performed a deep-copy, the `copyup` step will incorrectly overwrite the HEAD revision.

**Comment by trociny** [comment]
@dillaman I have no a good idea if it is related, but I was observing this fsx test failure [1] when running on a branch with this PR and #37581:
```
2020-10-11T16:12:52.307 INFO:teuthology.orchestra.run.smithi146.stdout:11091 read	0x67de519 thru	0x67ee060	(0xfb48 bytes)
2020-10-11T16:12:52.307 INFO:teuthology.orchestra.run.smithi146.stdout:short read: 0xe061 bytes instead of 0xfb48
```

[1]  http://qa-proxy.ceph.com/teuthology/trociny-2020-10-10_16:53:09-rbd-wip-mgolub-testing-3-distro-basic-smithi/5513900/teuthology.log

**Comment by dillaman** [comment]
It's a very confusing failure case. The IO was for extent `108913945~64328` and given the striping of image `image_client.0-clone70` (object-size 524288, stripe-unit 65536, stripe-count 6), the request would have been split between two objects:

`rbd_data.113034d5f896.00000000000000cc`:  `327680~57441`
`rbd_data.113034d5f896.00000000000000d1`: `320793~6887`

Given that the read was short by 6887 bytes, I'm presuming there was a failure in the read or processing against `rbd_data.113034d5f896.00000000000000d1`. Regardless of what the OSD returns, librbd will always invoke `AioCompletion::complete_request(object_extent_length)` which implies there is most likely an issue in `AioCompletion`. Digging into that function, it turns out that the master-branch which to multithreaded IO has exposed a race where the `AioCompletion` can finalize while there is still another thread inside `AioCompletion::complete_request` but before it had a chance to update the `rval`.

TL;DR: the failure doesn't appear to be related to either PR and I'll open a PR to address the race condition on the master branch.

**Comment by trociny** [comment]
@dillaman It needs rebase after #37581 has been merged.

**Comment by dillaman** [comment]
rebased

**Review by trociny — APPROVED** [review]
LGTM

**Comment by trociny** [comment]
@dillaman I run all your 3 open PRs on teuthology [1, 2]. There are some failures, the most interesting look [3, 4], though I have no idea if they related ([3] was a crash in journaler, [4] was qemu crash and the backtrace is printed without symbols).

So I think your PRs are good to be merged. I just want your confirmation before merging, just in case...

[1] https://pulpito.ceph.com/trociny-2020-10-13_17:35:53-rbd-wip-mgolub-testing-3-distro-basic-smithi/
[2] https://pulpito.ceph.com/trociny-2020-10-14_10:03:51-rbd-wip-mgolub-testing-3-distro-basic-smithi/
[3] http://qa-proxy.ceph.com/teuthology/trociny-2020-10-13_17:35:53-rbd-wip-mgolub-testing-3-distro-basic-smithi/5522911/teuthology.log
[4] http://qa-proxy.ceph.com/teuthology/trociny-2020-10-14_10:03:51-rbd-wip-mgolub-testing-3-distro-basic-smithi/5525677/teuthology.log

**Comment by dillaman** [comment]
It looks like the `qemu` crash was the same:

```
#0  0x00007f46e58908df in raise () from /lib64/libc.so.6
#1  0x00007f46e587acf5 in abort () from /lib64/libc.so.6
#2  0x00007f46cf3b5877 in ceph::__ceph_assert_fail (assertion=<optimized out>, file=<optimized out>, line=<optimized out>, 
    func=0x7f46dd5e3980 <journal::ObjectRecorder::notify_handler_unlock(std::unique_lock<std::mutex>&, bool)::__PRETTY_FUNCTION__> "void journal::ObjectRecorder::notify_handler_unlock(std::unique_lock<std::mutex>&, bool)")
    at /usr/src/debug/ceph-16.0.0-6312.gc1613349.el8.x86_64/src/common/assert.cc:75
#3  0x00007f46cf3b5a40 in ceph::__ceph_assert_fail (ctx=...) at /usr/src/debug/ceph-16.0.0-6312.gc1613349.el8.x86_64/src/common/assert.cc:80
#4  0x00007f46dd4acf7e in journal::ObjectRecorder::notify_handler_unlock (this=<optimized out>, locker=..., notify_overflowed=<optimized out>) at /usr/src/debug/ceph-16.0.0-6312.gc1613349.el8.x86_64/src/log/Entry.h:35
#5  0x00007f46dd4b2b69 in journal::ObjectRecorder::handle_append_flushed (this=0x7f4570002b10, tid=<optimized out>, r=<optimized out>) at /usr/src/debug/ceph-16.0.0-6312.gc1613349.el8.x86_64/src/journal/ObjectRecorder.cc:239
#6  0x00007f46dd4b3628 in Context::complete (r=<optimized out>, this=0x7f46883902b0) at /usr/src/debug/ceph-16.0.0-6312.gc1613349.el8.x86_64/src/include/Context.h:99
```
